discovery-aws-api-async/src/main/scala/akka/discovery/awsapi/ecs/AsyncEcsTaskSetDiscovery.scala

```diff
@@ -0,0 +1,217 @@
+
+  private[this] lazy val ecsClient = {
+    val conf = ClientOverrideConfiguration.builder().retryPolicy(RetryPolicy.none).build()
+    EcsAsyncClient.builder().overrideConfiguration(conf).build()
+  }
+
+  private[this] implicit val actorSystem: ActorSystem = system
+  private[this] implicit val materializer: ActorMaterializer = ActorMaterializer()
+  private[this] implicit val ec: ExecutionContext = system.dispatcher
+
+  private[this] val httpClient: HttpExt = Http()
+
+  override def lookup(lookup: Lookup, resolveTimeout: FiniteDuration): Future[Resolved] =
+    Future.firstCompletedOf(
+      Seq(
+        after(resolveTimeout, using = system.scheduler)(
+          Future.failed(new TimeoutException("Future timed out!"))
+        ),
+        resolveTasks(ecsClient, cluster, httpClient).map(
+          tasks =>
+            Resolved(
+              serviceName = lookup.serviceName,
+              addresses = for {
+                task <- tasks
+                container <- task.containers().asScala
+                networkInterface <- container.networkInterfaces().asScala
+              } yield {
+                val address = networkInterface.privateIpv4Address()
+                ResolvedTarget(host = address, port = None, address = Try(InetAddress.getByName(address)).toOption)
+              }
+            )
+        )
+      )
+    )
+
+}
+
+@ApiMayChange
+object AsyncEcsTaskSetDiscovery {
+
+  private[this] case class TaskMetadata(TaskARN: String)
+  private[this] case class TaskSet(value: String) extends AnyVal
+
+  private[this] implicit val orderFormat = jsonFormat1(TaskMetadata)
+
+  private val ECS_CONTAINER_METADATA_URI_PATH = "ECS_CONTAINER_METADATA_URI"
+
+  // InetAddress.getLocalHost.getHostAddress throws an exception when running
+  // in awsvpc mode because the container name cannot be resolved.
+  // ECS provides a metadata file
+  // (https://docs.aws.amazon.com/AmazonECS/latest/developerguide/container-metadata.html)
+  // that we ought to be able to use instead to find our IP address, but the
+  // metadata file does not get set when running on Fargate. So this is our
+  // only option for determining what the canonical Akka and akka-management
+  // hostname values should be set to.
+  def getContainerAddress: Either[String, InetAddress] =
+    NetworkInterface.getNetworkInterfaces.asScala
+      .flatMap(_.getInetAddresses.asScala)
+      .filterNot(_.isLoopbackAddress)
+      .filter(_.isSiteLocalAddress)
+      .toList match {
+      case List(value) =>
+        Right(value)
+
+      case other =>
+        Left(s"Exactly one private address must be configured (found: $other).")
+    }
+
+  private def resolveTasks(ecsClient: EcsAsyncClient, cluster: String, httpClient: HttpExt)(
+      implicit
+      ec: ExecutionContext,
+      mat: ActorMaterializer
+  ): Future[Seq[Task]] =
+    for {
+      taskArn <- resolveTaskMetadata(httpClient).map(_.map(_.TaskARN))
+      taskSet <- taskArn match {
```

Comment:
Didn't want to pull in cats to have `OptionT`, so I came up with this. Suggestions are welcome to make this cleaner.